source/extensions/tracers/skywalking/trace_segment_reporter.cc

```diff
@@ -0,0 +1,169 @@
+    // get more information.
+    span->set_componentid(9000);
+
+    if (!span_store->peerAddress().empty() && span_store->isEntrySpan()) {
+      span->set_peer(span_store->peerAddress());
+    }
+
+    span->set_spanid(span_store->spanId());
+    span->set_parentspanid(span_store->parentSpanId());
+
+    span->set_starttime(span_store->startTime());
+    span->set_endtime(span_store->endTime());
+
+    span->set_iserror(span_store->isError());
+
+    span->set_operationname(span_store->operation());
+
+    auto& tags = *span->mutable_tags();
+    tags.Reserve(span_store->tags().size());
+
+    for (auto& span_tag : span_store->tags()) {
+      KeyStringValuePair* new_tag = tags.Add();
+      new_tag->set_key(span_tag.first);
+      new_tag->set_value(span_tag.second);
+    }
+
+    SpanContext* previous_span_context = segment_context.previousSpanContext();
+
+    if (!previous_span_context || !span_store->isEntrySpan()) {
+      continue;
+    }
+
+    auto* ref = span->mutable_refs()->Add();
+    ref->set_traceid(previous_span_context->trace_id_);
+    ref->set_parenttracesegmentid(previous_span_context->trace_segment_id_);
+    ref->set_parentspanid(previous_span_context->span_id_);
+    ref->set_parentservice(previous_span_context->service_);
+    ref->set_parentserviceinstance(previous_span_context->service_instance_);
+    ref->set_parentendpoint(previous_span_context->endpoint_);
+    ref->set_networkaddressusedatpeer(previous_span_context->target_address_);
+  }
+  return new_segment_ptr;
+}
+
+} // namespace
+
+TraceSegmentReporter::TraceSegmentReporter(Grpc::AsyncClientFactoryPtr&& factory,
+                                           Event::Dispatcher& dispatcher,
+                                           Random::RandomGenerator& random_generator,
+                                           SkyWalkingTracerStats& stats,
+                                           const SkyWalkingClientConfig& client_config)
+    : tracing_stats_(stats), client_config_(client_config), client_(factory->create()),
+      service_method_(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+          "TraceSegmentReportService.collect")),
+      random_generator_(random_generator) {
+
+  static constexpr uint32_t RetryInitialDelayMs = 500;
+  static constexpr uint32_t RetryMaxDelayMs = 30000;
+  backoff_strategy_ = std::make_unique<JitteredExponentialBackOffStrategy>(
+      RetryInitialDelayMs, RetryMaxDelayMs, random_generator_);
+
+  retry_timer_ = dispatcher.createTimer([this]() -> void { establishNewStream(); });
+  establishNewStream();
+}
+
+void TraceSegmentReporter::onCreateInitialMetadata(Http::RequestHeaderMap& metadata) {
+  if (!client_config_.backendToken().empty()) {
+    metadata.setReferenceKey(authenticationTokenKey(), client_config_.backendToken());
+  }
+}
+
+void TraceSegmentReporter::report(const SegmentContext& segment_context) {
+  sendTraceSegment(toSegmentObject(segment_context));
```

Comment:
I think that `TraceSegmentPtr` can be moved. So we shouldn't use `TraceSegmentPtr&&` but `TraceSegmentPtr`.